File: datastructure/04string/code/最长回文子串.java
class Solution {
    public String longestPalindrome(String s) {
        int l=0,r=0,max=1;
        for(int i=0;i<s.length();i++){
            int len=1;
            for(int k=1;i-k>=0&&i+k<s.length();k++){
                if(s.charAt(i-k)==s.charAt(i+k)){
                    len+=2;
                    if(len>max){
                        l=i-k;
                        r=i+k;
                        max=len;
                    }
                }else break;
            }
            len=1;
        }
        for(int i=1;i<s.length();i++){
            int len=0;
            for(int k=1;i-k>=0&&i+k-1<s.length();k++){
                if(s.charAt(i-k)==s.charAt(i+k-1)){
                    len+=2;
                    if(len>max){
                        l=i-k;
                        r=i+k-1;
                        max=len;
                    }
                }
                else break;
            }
            len=0;
        }
        return s.substring(l,r+1);
    }
}

File: algorithm/dp/02区间dp/code/最长回文子串.cpp
class Solution {
public:
// https://leetcode.cn/problems/longest-palindromic-substring/
    string longestPalindrome(string s) {
        int n = s.size();
        bool dp[n][n];
        memset(dp, 0, sizeof(dp));
        for (int i = 0; i < n; i++) {
            dp[i][i] = true;
        }
        string ans = s.size() == 0 ? "" : s.substr(0, 1);
        for (int len = 2; len <= n; len++) {
            for (int i = 0; i + len - 1 < n; i++) {
                int left = i, right = i + len - 1;
                if (left == right - 1) {
                    dp[left][right] = s[left] == s[right];
                } else {
                    dp[left][right] = dp[left + 1][right - 1] && s[left] == s[right];
                }
                if (dp[left][right] && len > ans.size()) {
                    ans = s.substr(left, len);
                }
            }
        }

        return ans;
    }
};